### ISSUE TITLE
SCOPE

### ISSUE THREAD

**Issue by rujinwang** [Bioconductor/Contributions#1242]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/rujinwang/SCOPE

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by rujinwang** [comment]
AdditionalPackage: https://github.com/rujinwang/WGSmapp

**Comment by rujinwang** [comment]
AdditionalPackage: https://github.com/rujinwang/WGSmapp

**Comment by rujinwang** [comment]
Dear Bioconductor Core Member @hpages ,
I would like this repository to be linked to issue number #1239. The software package SCOPE has the companion experiment data package WGSmapp used for illustrative purposes in the vignette. Otherwise the build of SCOPE will get ERRORs. Could you please help me on this? Thanks.

Rujin

**Comment by hpages** [comment]
Hi Rujin,

Thanks for your submission.

You don't need to open a new issue for companion data packages so please close issue #1239.

All you need to do is add the companion package as an additional package to the issue you already opened for the software package (i.e. this issue). Which is what you did 2 days ago (see https://github.com/Bioconductor/Contributions/issues/1242#issuecomment-529566331).

Looks like our Single Package Builder (SPB) failed to install WGSmapp before trying to run `R CMD build` on SCOPE. Not sure why. Maybe @lshep can chime in. But before that, please close issue #1239 and bump SCOPE version (to 0.99.1) to trigger a new build by the SPB and we'll see what happens.

Thanks,
H.

**Comment by hpages** [comment]
> Do I need to use devtools::install_github("rujinwang/WGSmapp") instead?

No because it's considered bad practice to install packages in a vignette. And having a vignette that installs stuff from GitHub is even worse.

I guess it's time to ask help from @lshep . Lori?

Thanks!

**Comment by hpages** [comment]
Lori (@lshep) contacted me privately to let me know that she'll take care of this tomorrow. Thanks for your patience.

**Comment by rujinwang** [comment]
Sounds good. Thank you for your help!

**Comment by hpages** [comment]
Hi @rujinwang ,

I see you've been actively working on SCOPE and the package now passes `R CMD check` and `R CMD BiocCheck` without errors or warnings. Thanks for that. Please let me know if the package is now ready for review.

Best,
H.

**Comment by rujinwang** [comment]
Hi @hpages ,
Yes, please go ahead to review the package. Thanks!

Best, 
Rujin

**Comment by hpages** [comment]
Hi Rujin,

I've taken a first look at the package and it has usability and reliability issues that are too serious for inclusion in Bioconductor. I bumped into these issues at the very beginning of your workflow, only after using the first two functions used in the workflow (`getbambed_scope` and `getmapp`). I could keep going e.g. `getgc_scope` has issues similar to `getmapp` and so on... (and why would one function be suffixed with `_scope` and not the other one?) but I'll stop here because my time is limited. If you want to pursue this submission, a lot of work will need to be done on the package to make it more usable and more reliable.

Regards,
H.

#### Some examples of usability issues

- The "1.2 Bioinformatic pre-processing" section of the vignette shows the use of command-line tools `java -jar picard.jar` and `split_script.py`. Are these tools available somewhere? Are they documented? What do they do exactly? If the user doesn't have access to them, what is the point of this section?

- Your function naming style is very inconsistent, even for the user-facing functions: you use camel caps (`getGini`), snake case (`getgc_scope`), dot-separated (`getcoverage.scDNA`), other unidentified styles (e.g. `choiceofT`). Please choose one style and stick to it. Also, a name like `getbambed_scope` is poor style. It's considered good practice to separate _all_ the words, not just _some_ of them. So a better name would be something like `get_bam_bed_scope` (snake case) or `getBamBedScope` (camel caps).

- The `getbambed_scope` function is poorly implemented. If you want to generate bins on a given genome, use `tileGenome()` from the GenomicRanges package:
    ```
    bins <- tileGenome(seqinfo(BSgenome.Hsapiens.UCSC.hg19), tilewidth=500000, cut.last.tile.in.chrom=TRUE)
    bins
    # GRanges object with 6343 ranges and 0 metadata columns:
    #                seqnames          ranges strand
    #                   <Rle>       <IRanges>  <Rle>
    #      [1]           chr1        1-500000      *
    #      [2]           chr1  500001-1000000      *
    #      [3]           chr1 1000001-1500000      *
    #      [4]           chr1 1500001-2000000      *
    #      [5]           chr1 2000001-2500000      *
    #      ...            ...             ...    ...
    #   [6339] chrUn_gl000245         1-36651      *
    #   [6340] chrUn_gl000246         1-38154      *
    #   [6341] chrUn_gl000247         1-36422      *
    #   [6342] chrUn_gl000248         1-39786      *
    #   [6343] chrUn_gl000249         1-38502      *
    #   -------
    #   seqinfo: 93 sequences (1 circular) from hg19 genome
    ```
    If `tileGenome()` doesn't cover you needs, you are welcome to provide your own function (could be a wrapper around `tileGenome()`) but please note that: (1) using a precomputed `.bed` file for that isn't necessary and is a bad idea, and (2) creating one bin at a time in a `while`-loop is also a bad idea because it is very inefficient.

- Having functions defined in your package referencing variables defined in the global environment is bad software design:
    ```
    getmapp(ref_raw, genome=BSgenome.Hsapiens.UCSC.hg38)
    # Error in getmapp(ref_raw, genome = BSgenome.Hsapiens.UCSC.hg38) : 
    #   object 'mapp_hg38' not found
    ```

- The design of the `getmapp()` function makes no sense at all. The user is required to pass a huge BSgenome object to it only to allow the function to know which mappability track to pick up. Why not just pass the mappability track instead? Then you wouldn't have to pick up the mappability track from the global environment. Another advantage to this design is that `getmapp()` then would work for an arbitrary genome, not just hg19 or hg38, as long as the user has access to the mappability track for the genome.

#### Some examples of reliability issues

- Do you realize that the BAM files you use in the "2.1 Pre-preparation" section of the vignette contain alignments against genome assembly hg38 so when you do
    ```
    bambedObj <- getbambed_scope(bamdir = bamdir, sampname = sampname_raw)
    ```
    without specifying `hgref="hg38"` you get the wrong binning? Also later when you do
    ```
    data("mapp_hg19")
    mapp <- getmapp(ref_raw, genome = BSgenome.Hsapiens.UCSC.hg19)
    ```
    you get the wrong mappability?

- This result is incorrect:
    ```
    getmapp(GRanges("chr1:10022-10031"), BSgenome.Hsapiens.UCSC.hg19)
    # Getting mappability for chr1
    # [1] 0.65625
    ```
    Should be 0.75 (half of the range has a mappability score of 0.5 and the other half a mappability score of 1).

#### Some other minor issues

- Add DNASeq to the biocViews field.

- Dataset `mapp_hg19` contains the chromosome lengths but not dataset `mapp_hg38`:
    ```
    seqlengths(mapp_hg19)
    #      chr1     chr10     chr11     chr12     chr13     chr14     chr15     chr16 
    # 249250621 135534747 135006516 133851895 115169878 107349540 102531392  90354753 
    #     chr17     chr18     chr19      chr2     chr20     chr21     chr22      chr3 
    #  81195210  78077248  59128983 243199373  63025520  48129895  51304566 198022430 
    #      chr4      chr5      chr6      chr7      chr8      chr9      chrM      chrX 
    # 191154276 180915260 171115067 159138663 146364022 141213431     16571 155270560 
    #      chrY 
    #  59373566
 
    seqlengths(mapp_hg38)
    #  chr1 chr20 chr21 chr15  chr8 chr10  chr9  chrX  chr3 chr12  chr2 chr11  chr7 
    #    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA 
    # chr13 chr18  chr5 chr22 chr14  chr4 chr17 chr16  chr6 chr19  chrM  chrY 
    #    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA    NA 
    ```
    Any reason for that?

- Coding style: you sometimes use `<-` and sometimes `=` for assigment. Please always use `<-` for assignment (with a space before and after the left arrow).

**Comment by rujinwang** [comment]
Hi @hpages , 
Thanks a lot for your review and constructive suggestions. I've made changes accordingly and bumped the package to a new version. Regarding some of your concerns, please see my responses below. 

**Some examples of usability issues**

> The "1.2 Bioinformatic pre-processing" section of the vignette shows the use of command-line tools java -jar picard.jar and split_script.py. Are these tools available somewhere? Are they documented? What do they do exactly? If the user doesn't have access to them, what is the point of this section?

The picard.jar is a public available tool for manipulating high-throughput sequencing data, released by Broad Institute. The only self-developed script is split_script.py and now it is pre-stored in the package (inst/docs folder). The aim of this section is to show how the pre-processing bioinformatic pipeline works for users to get preparation of the input for SCOPE (demultiplexed .bam files). 

- Functions have been renamed using snake case. Now the naming style is consistent. 

- For the design of function `get_mapp()`, it aims to compute mappability for each bin. `mapp_hg38` or `mapp_hg19` are large-size GRanges Objects pre-stored in the dependent dataset package `WGSmapp`, rather than variables defined in the global environment. We developed the dataset package `WGSmapp` in particular to offer mappability tracks of whole-genome sequencing. 
```
library(WGSmapp)
data("mapp_hg38")
getmapp(ref_raw, hgref = "hg38")
```

These pre-stored mappability tracks are downloaded from the [ENCODE Project](http://hgdownload.cse.ucsc.edu/goldenpath/hg19/encodeDCC/wgEncodeMapability/) thus can not be directly passed to a function. I'm now passing a simple character indicator, instead of a huge BSgenome object, to allow the function to know which mappability track to pick up. Do you have any comments or suggestions on how to do it in a more efficient way? 

**Some examples of reliability issues**

> Do you realize that the BAM files you use in the "2.1 Pre-preparation" section of the vignette contain alignments against genome assembly hg38

How did you realize these BAM files are against genome assembly hg38? I downloaded them from 10XGenomic website, which is public available. Given the information provided [here](http://cf.10xgenomics.com/samples/cell-dna/1.1.0/bj_mkn45_10pct/bj_mkn45_10pct_web_summary.html), it's against assembly GRCh37/hg19. 


> This result is incorrect:
> ```
> getmapp(GRanges("chr1:10022-10031"), BSgenome.Hsapiens.UCSC.hg19)
> # Getting mappability for chr1
> # [1] 0.65625
> ```
> Should be 0.75 (half of the range has a mappability score of 0.5 and the other half a mappability score of 1).

The mappability is computed as weighted average of the mappability scores if multiple ENCODE regions overlap with the query genome region. It's not weighted by the width of the query genome region, but rather by the width of mappability track bins. In other word, ranges that only hit regions with mappability scores of 0.5 and 1 will have the same mappability scores, whatever the width of each sub-region is. For example, in your scenario, mappability = (0.5*11+1*5)/(11+5) = 0.65625.


**Some other minor issues**
- DNASeq is added to the biocViews 
- Both `mapp_hg19` and `mapp_hg38` contain chromosome lengths now
- Coding style for assignment is consistent now

Yes, we are pursuing the submission. Please feel free to let me know if further modifications or improvements are needed. Thank you for your help!

Best, 
Rujin

**Comment by lshep** [comment]
Please don't be alarmed about all the build report.  And sorry for the noise.  We are updating the builders to R 4.0 and Bioc 3.11 -  Your package has some special cases that are triggering ERRORs on our end in windows -  Unfortunately we don't have a good test case to run separately besides kicking off new builds of your package -   Again we apologize for the noise and hope to have it remedied soon.

**Comment by hpages** [comment]
Hi Rujin,

Thanks for the improvements to the package.

#### Previously discussed

> `mapp_hg38` or `mapp_hg19` are large-size GRanges Objects pre-stored in the dependent dataset package WGSmapp, rather than variables defined in the global environment.

But when you do `data("mapp_hg38")` you load the `mapp_hg38` dataset in the global environment. So now you have a variable `mapp_hg38` defined in the global environment. And your `get_mapp()` function expects this variable to be there. If the variable is not there, `get_mapp()` won't work (it will fail).
This is fragile. This is why having a function defined in your package referencing variables defined in the global environment is considered bad software design. In other words, `get_mapp()` should always be able to find the `mapp_hg38` dataset (whether the user previously did `data("mapp_hg38")` or not).

> These pre-stored mappability tracks are downloaded from the ENCODE Project thus can not be directly passed to a function.

Why can't they? You've saved them as GRanges objects. Why couldn't you pass a GRanges object to a function? E.g.:
```
data("mapp_hg38")
get_mapp(ref, mapp_hg38)
```
As easy as that!

> How did you realize these BAM files are against genome assembly hg38?

I looked at the chromosome lengths:
```
> seqinfo(BamFile(bambedObj$bamdir[[1]]))
Seqinfo object with 196 sequences from an unspecified genome:
  seqnames         seqlengths isCircular genome
  chr1              248956422       <NA>   <NA>
  chr2              242193529       <NA>   <NA>
  chr3              198295559       <NA>   <NA>
  chr4              190214555       <NA>   <NA>
  chr5              181538259       <NA>   <NA>
  ...                     ...        ...    ...
  chrUn_KI270742v1     186739       <NA>   <NA>
  chrUn_GL000216v2     176608       <NA>   <NA>
  chrUn_GL000218v1     161147       <NA>   <NA>
  chrEBV               171823       <NA>   <NA>
  hs38d1             10560522       <NA>   <NA>
```
Those are the chromosome lengths of hg38 (see `seqinfo(getBSgenome("hg38"))`), not hg19.

> It's not weighted by the width of the query genome region, but rather by the width of mappability track bins.

So if you have a range of 1 million bases where only the 1st position overlaps with a bin of mappability 0.5 and the 999999 other positions overlap with a bin of mappability 1, and the two bins have the same size, you're considering that the mappability of your range is 0.75? Even though 99.9999% of the range has a mappability of 1? How does that make sense?

#### Other things

- You're not making good use of argument default values. For example why set the default value for the `hgref` argument to `NULL` (in `get_bam_bed()` and `get_mapp()`) when the **effective** default is `"hg19"`? Or why set the default value for the `resolution` argument (in `get_bam_bed()`) to `NULL` when the **effective** default is 500?

- Don't put an exclamation mark (`!`) at the end of your error messages.

Best,
H.

**Comment by rujinwang** [comment]
Hi @hpages ,
Thanks a lot for your comments and constructive suggestions. I have been working on this paper revision and it took a bit longer. I've made changes accordingly and bumped the package to a new version. 
To be specific, 
**Previously discussed**
> You've saved them as GRanges objects. Why couldn't you pass a GRanges object to a function?
Yes, thanks for your suggestion! The current version does not load pre-stored dataset in the global environment.

> Those are the chromosome lengths of hg38 (see `seqinfo(getBSgenome("hg38"))`), not hg19.
Yes, you are correct. I loaded the bam files into IGV and confirmed it's hg38. I made changes in function examples by specifying their reference genome to be hg38. 

> So if you have a range of 1 million bases where only the 1st position overlaps with a bin of mappability 0.5 and the 999999 other positions overlap with a bin of mappability 1, and the two bins have the same size, you're considering that the mappability of your range is 0.75? Even though 99.9999% of the range has a mappability of 1? How does that make sense?
Regarding mappability calculation, I agreed with you that the one below would make more sense. `get_mapp()` has been adjusted. 
- (mapp_A*size_overlap_with_A + mapp_B*size_overlap_with_B)/(size_overlap_with_A+size_overlap_with_B)

**Other things**
- Argument default values with `NULL` have been reset. `get_bam_bed()`, `get_mapp()` and `get_gc` have been updated. 
- The exclamation marks at the end of your error messages have been removed. 

Please feel free to let me know if further improvements are needed. Thank you for your kind help!

Best regards, 
Rujin

**Comment by rujinwang** [comment]
Hi @hpages ,
Just want to check in on it. I have been working on improving the package. Feel free to let me know your comments. Thank you. 

Best, 
Rujin

**Comment by hpages** [comment]
Thanks for letting me know. Will take a look ASAP.

**Comment by hpages** [comment]
Thanks @rujinwang. Package is good to go.

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/rujinwang.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("SCOPE")`. The package 'landing page' will be created at

https://bioconductor.org/packages/SCOPE

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/

**Comment by mtmorgan** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/rujinwang.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("WGSmapp")`. The package 'landing page' will be created at

https://bioconductor.org/packages/WGSmapp

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
